cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java

```diff
@@ -138,4 +316,470 @@ private String getProgramScheduleId(Id.Program program, SchedulableProgramType p
+        MetricType.COUNTER,
+        ImmutableMap.of(Constants.Metrics.Tag.NAMESPACE, streamId.getNamespaceId(),
+                        Constants.Metrics.Tag.STREAM, streamId.getName()),
+        ImmutableList.<String>of()
+      );
+
+      try {
+        Collection<MetricTimeSeries> metrics = metricStore.query(metricDataQuery);
+        if (metrics == null || metrics.isEmpty()) {
+          // Data is not yet available, which means no data has been ingested by the stream yet
+          return new StreamSize(0L, System.currentTimeMillis());
+        }
+
+        MetricTimeSeries metric = metrics.iterator().next();
+        List<TimeValue> timeValues = metric.getTimeValues();
+        if (timeValues == null || timeValues.size() != 1) {
+          throw new IOException("Should collect exactly one time value");
+        }
+        TimeValue timeValue = timeValues.get(0);
+        // The metric store gives us 0 as the timestamp, hence we cannot use it here
+        return new StreamSize(timeValue.getValue(), System.currentTimeMillis());
+      } catch (Exception e) {
+        Throwables.propagateIfInstanceOf(e, IOException.class);
+        throw new IOException(e);
+      }
+    }
+  }
+
+  /**
+   * Wrapper around a {@link StreamSizeSchedule} which will run a program whenever it receives enough
+   * data from a stream, via notifications.
+   */
+  private final class StreamSizeScheduleTask {
+    private final Id.Program programId;
+    private final SchedulableProgramType programType;
+    private StreamSizeSchedule streamSizeSchedule;
+
+    private long baseSize;
+    private long baseTs;
+    private AtomicBoolean active;
+
+    private StreamSizeScheduleTask(Id.Program programId, SchedulableProgramType programType,
+                                   StreamSizeSchedule streamSizeSchedule) {
+      this.programId = programId;
+      this.programType = programType;
+      this.streamSizeSchedule = streamSizeSchedule;
+    }
+
+    public void startSchedule(long baseSize, long baseTs, boolean active, boolean persist) {
+      LOG.debug("Starting schedule {} with baseSize {}, baseTs {}, active {}. Should be persisted: {}",
+                streamSizeSchedule.getName(), baseSize, baseTs, active, persist);
+      this.baseSize = baseSize;
+      this.baseTs = baseTs;
+      this.active = new AtomicBoolean(active);
+    }
+
+    public boolean isActive() {
+      return active.get();
+    }
+
+    public void received(StreamSizeNotification notification) {
+      if (!active.get()) {
+        return;
+      }
+      long pastRunSize;
+      long pastRunTs;
+      StreamSizeSchedule currentSchedule;
+      synchronized (this) {
+        if (notification.getSize() < baseSize) {
+          // This can happen if no notification is received for a stream for some time, and we poll the stream events
+          // size using metrics, and some metric events have hit their TTL. In that case it's impossible to know
```

Comment:
For your last comment, if there was a race, then the notification containing the polling would have a timestamp lower than the last notification. But the received method here would never be called with such a notification - we make sure of that in the StreamSubscriber#received. Yet, now that I think of it, the polling could indeed show a timestamp higher - ie more recent - but a lower size, because the metrics shows a state which could be a few moments earlier than the timestamp we attach it with. This is because the way we query metrics, we can't retrieve the timestamp of the last seen event - I checked with @gsps1, but will make sure again. In that situation, I don't know what should be done. The better option would be to open a Jira for the metrics issue, like you suggested below. @gsps1 told me that it was supposed to be that way, though.